### ISSUE TITLE
[Bug]: Segmentation fault in PyTorch tensor deallocation during multiprocessing training with ManiSkill environment

### ISSUE THREAD

**Issue by zixiao-bios** [RLinf/RLinf#131]
### Affected component

Environment Management, Multiprocessing Workers, CUDA/GPU Integration

### Bug description

I'm not familiar with core dump analysis, and the core dump is too big to upload. I asked AI to analyze the core dump, and the following are its results (Just tell me if you need other information):

#### Summary
Training runs successfully with continuous TensorBoard updates, but generates multiple core dump files due to segmentation faults occurring in PyTorch tensor deallocation during multiprocessing.

#### Core Dump Analysis
Based on GDB analysis of the core dump files, the crash occurs in:
Program terminated with signal SIGSEGV, Segmentation fault.
#0 Py_TYPE (ob=ob@entry=0x1) at /usr/local/src/conda/python-3.11.10/Include/object.h:133
#6 0x00007f14c7658f53 in THPVariable_subclass_dealloc(object) ()
from /opt/conda/lib/python3.11/site-packages/torch/lib/libtorch_python.so
The crash happens in `THPVariable_subclass_dealloc` when trying to access invalid memory address `0x1`, indicating a memory corruption or double-free issue in PyTorch tensor cleanup.

#### Key Observations
1. **Training continues normally**: The main training process keeps running and TensorBoard shows continuous updates
2. **Multiple core dumps**: 16 core dump files (~5.5GB each) generated during training
3. **Multiprocessing context**: Crash occurs in child processes spawned via `multiprocessing.spawn`
4. **GPU backend**: Using ManiSkill with `sim_backend: "gpu"` and `enable_offload: True`

#### Suspected Root Causes
1. **NUMA affinity binding issues** in `rlinf/envs/env_manager.py:set_process_numa_affinity()`
2. **CUDA tensor sharing** between processes in offload wrapper
3. **GPU memory management** during environment state serialization/deserialization
4. **Race conditions** in `_simulator_worker` process cleanup

### Environment

Python 3.11.10
arm_pytorch_utilities              0.4.3
array_record                       0.8.1
pytorch-kinematics                 0.7.5
pytorch-seed                       0.2.0
ray                                2.47.0
torch                              2.5.1+cu124
torch_memory_saver                 0.0.8
torchaudio                         2.5.1+cu124
torchdata                          0.11.0
torchelastic                       0.2.2
torchvision                        0.20.1+cu124
transformers                       4.40.1
Wed Sep 24 16:49:58 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 570.124.06             Driver Version: 570.124.06     CUDA Version: 12.8     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA H100 80GB HBM3          On  |   00000000:18:00.0 Off |                    0 |
| N/A   27C    P0             69W /  700W |       1MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA H100 80GB HBM3          On  |   00000000:2A:00.0 Off |                    0 |
| N/A   28C    P0             74W /  700W |       1MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   2  NVIDIA H100 80GB HBM3          On  |   00000000:3A:00.0 Off |                    0 |
| N/A   29C    P0             69W /  700W |       1MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   3  NVIDIA H100 80GB HBM3          On  |   00000000:5D:00.0 Off |                    0 |
| N/A   28C    P0             70W /  700W |       1MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   4  NVIDIA H100 80GB HBM3          On  |   00000000:84:00.0 Off |                    0 |
| N/A   27C    P0             69W /  700W |       1MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   5  NVIDIA H100 80GB HBM3          On  |   00000000:8B:00.0 Off |                    0 |
| N/A   29C    P0             69W /  700W |       1MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   6  NVIDIA H100 80GB HBM3          On  |   00000000:91:00.0 Off |                    0 |
| N/A   28C    P0             69W /  700W |       1MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   7  NVIDIA H100 80GB HBM3          On  |   00000000:E4:00.0 Off |                    0 |
| N/A   27C    P0             69W /  700W |       1MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+

### Reproduction

My docker image: `rlinf/rlinf:agentic-openvla-rlinf0.1-torch2.5.1`
My config file:
```json
{
    "env": {
        "train": {
            "simulator_type": "maniskill",
            "auto_reset": true,
            "ignore_terminations": true,
            "use_rel_reward": true,
            "seed": 0,
            "num_group": 15,
            "group_size": 1,
            "use_fixed_reset_state_ids": false,
            "only_eval": false,
            "max_episode_steps": 80,
            "video_cfg": {
                "save_video": true,
                "info_on_video": true,
                "video_base_dir": "RLinf/logs/20250924-13:04:37_maniskill_ppo_openvla_V0.2/video/train"
            },
            "init_params": {
                "id": "PutOnPlateInScene25Main-v3",
                "num_envs": 15,
                "obs_mode": "rgb+segmentation",
                "control_mode": "arm_pd_ee_target_delta_pose_align2_gripper_pd_joint_pos",
                "sim_backend": "gpu",
                "sim_config": {
                    "sim_freq": 500,
                    "control_freq": 5
                },
                "max_episode_steps": 80,
                "sensor_configs": {
                    "shader_pack": "default"
                },
                "render_mode": "all",
                "obj_set": "train"
            }
        },
        "eval": {
            "simulator_type": "maniskill",
            "auto_reset": true,
            "ignore_terminations": true,
            "use_rel_reward": true,
            "seed": 0,
            "num_group": 1,
            "group_size": 1,
            "use_fixed_reset_state_ids": true,
            "num_envs": 1,
            "only_eval": true,
            "video_cfg": {
                "save_video": true,
                "info_on_video": true,
                "video_base_dir": "RLinf/logs/20250924-13:04:37_maniskill_ppo_openvla_V0.2/video/eval"
            },
            "init_params": {
                "id": "PutOnPlateInScene25Main-v3",
                "num_envs": 4,
                "obs_mode": 1,
                "control_mode": "arm_pd_ee_target_delta_pose_align2_gripper_pd_joint_pos",
                "sim_backend": "gpu",
                "sim_config": {
                    "sim_freq": 500,
                    "control_freq": 5
                },
                "max_episode_steps": 80,
                "sensor_configs": {
                    "shader_pack": "default"
                },
                "render_mode": "all",
                "obj_set": "test"
            }
        },
        "group_name": "EnvGroup",
        "channel": {
            "name": "env_buffer_list",
            "queue_name": "obs_buffer",
            "queue_size": 0
        },
        "enable_offload": true
    },
    "cluster": {
        "num_nodes": 1,
        "num_gpus_per_node": 8,
        "component_placement": {
            "actor": "0-7",
            "env": "0-3",
            "rollout": "4-7"
        }
    },
    "runner": {
        "task_type": "embodied",
        "logger": {
            "log_path": "RLinf/logs/20250924-13:04:37_maniskill_ppo_openvla_V0.2",
            "project_name": "rlinf",
            "experiment_name": "test_openvla",
            "logger_backends": [
                "tensorboard"
            ]
        },
        "max_epochs": 1000,
        "max_steps": -1,
        "only_eval": false,
        "val_check_interval": -1,
        "save_interval": 40,
        "seq_length": 4096,
        "max_prompt_length": 30
    },
    "algorithm": {
        "auto_reset": true,
        "ignore_terminations": true,
        "use_fixed_reset_state_ids": false,
        "require_values": true,
        "shuffle_samples": true,
        "normalize_advantages": true,
        "kl_penalty": "kl",
        "group_size": 1,
        "n_chunk_steps": 80,
        "n_eval_chunk_steps": 80,
        "num_group_envs": 15,
        "rollout_epoch": 1,
        "reward_type": "step_level",
        "logprob_type": "action_level",
        "entropy_type": "action_level",
        "adv_type": "embodied_gae",
        "loss_type": "embodied_ppo",
        "loss_agg_func": "token-mean",
        "kl_beta": 0.0,
        "entropy_bonus": 0,
        "clip_ratio_high": 0.2,
        "clip_ratio_low": 0.2,
        "clip_ratio_c": 3.0,
        "value_clip": 0.2,
        "huber_delta": 10.0,
        "gamma": 0.99,
        "gae_lambda": 0.95,
        "sampling_params": {
            "use_greedy": false,
            "temperature_train": 1.0,
            "temperature_eval": 0.6,
            "top_k": 50,
            "top_p": 1.0,
            "repetition_penalty": 1.0
        },
        "length_params": {
            "max_new_token": 7,
            "max_length": 1024,
            "min_length": 1
        }
    },
    "rollout": {
        "group_name": "RolloutGroup",
        "channel": {
            "name": "env_buffer_list",
            "queue_name": "action_buffer",
            "queue_size": 0
        },
        "mode": "colocate",
        "backend": "huggingface",
        "generation_backend": "huggingface",
        "model_dir": "/inspire/ssd/project/robotsimulation/public/eric/Models/openvla-7b-rlvla-warmup/",
        "enable_offload": true,
        "pipeline_stage_num": 2
    },
    "actor": {
        "group_name": "ActorGroup",
        "channel": {
            "name": "env_buffer_list",
            "queue_name": "replay_buffer",
            "queue_size": 0
        },
        "training_backend": "fsdp",
        "checkpoint_load_path": "/inspire/ssd/project/robotsimulation/public/eric/Models/openvla-7b-rlvla-warmup/",
        "checkpoint_save_path": "../results",
        "micro_batch_size": 20,
        "global_batch_size": 160,
        "seed": 1234,
        "enable_offload": true,
        "model": {
            "model_name": "openvla",
            "action_dim": 7,
            "num_action_chunks": 1,
            "use_proprio": false,
            "unnorm_key": "bridge_orig",
            "center_crop": true,
            "do_sample": false,
            "precision": "bf16",
            "add_bias_linear": false,
            "add_qkv_bias": true,
            "vocab_size": 32000,
            "hidden_size": 4096,
            "policy_setup": "widowx_bridge",
            "vh_mode": "a0",
            "image_size": [
                224,
                224
            ],
            "is_lora": true,
            "lora_rank": 32,
            "ckpt_path": null,
            "attn_implementation": "flash_attention_2",
            "low_cpu_mem_usage": true,
            "trust_remote_code": true,
            "sharding_strategy": "full_shard"
        },
        "optim": {
            "lr": 0.0001,
            "value_lr": 0.003,
            "adam_beta1": 0.9,
            "adam_beta2": 0.95,
            "adam_eps": 1e-05,
            "clip_grad": 1.0
        },
        "tokenizer": {
            "tokenizer_type": "HuggingFaceTokenizer",
            "tokenizer_model": "/inspire/ssd/project/robotsimulation/public/eric/Models/openvla-7b-rlvla-warmup/",
            "use_fast": false,
            "trust_remote_code": true,
            "padding_side": "right"
        }
    },
    "reward": {
        "use_reward_model": false
    },
    "critic": {
        "use_critic_model": false
    },
}
```

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by zixiao-bios** [comment]
I located the problem: **When the env instance is destroyed, it will cause a segment fault**. If setting env.enable_offload = True, the env instance is destroyed after each rollout, thus generating a large number of core dump files.
It seems not a RLinf Bug, may be related to Maniskill or Vulkan.
I simply set env.enable_offload = False to avoid it.
